Validation/HGCalValidation/test/HGCalWaferHitCheck.cc

```diff
@@ -0,0 +1,257 @@
+      analyzeHits(nameDetector_, *(theRecHitContainer.product()));
+    } else if (verbosity_ > 0) {
+      edm::LogVerbatim("HGCalValidation") << "RecHitContainer does not exist for " << nameDetector_;
+    }
+  }
+}
+
+template <class T>
+void HGCalWaferHitCheck::analyzeHits(std::string const& name, T const& hits) {
+  for (auto const& hit : hits) {
+    uint32_t id = hit.id();
+    int zside, type, layer, waferU, waferV;
+    if (ifNose_) {
+      HFNoseDetId detId = HFNoseDetId(id);
+      waferU = detId.waferU();
+      waferV = detId.waferV();
+      type = detId.type();
+      layer = detId.layer();
+      zside = detId.zside();
+    } else {
+      HGCSiliconDetId detId = HGCSiliconDetId(id);
+      waferU = detId.waferU();
+      waferV = detId.waferV();
+      type = detId.type();
+      layer = detId.layer();
+      zside = detId.zside();
+    }
+    int index = HGCalWaferIndex::waferIndex(layer, waferU, waferV, false);
+    int typef = hgcons_->waferType(layer, waferU, waferV, true);
+    if (zside < 0)
+      index *= -1;
+    auto itr = occupancy_.find(index);
+    if (itr == occupancy_.end())
+      occupancy_[index] = 1;
+    else
+      ++occupancy_[index];
+    if (type != typef) {
+      auto ktr = badTypes_.find(index);
+      if (ktr == badTypes_.end())
+        badTypes_[index] = 1;
+      else
+        ++badTypes_[index];
+      if (verbosity_ == 1)
+        edm::LogVerbatim("HGCalValidation")
+            << "Detector " << name << " zside = " << zside << " layer = " << layer << " type = " << type << ":" << typef
+            << " wafer = " << waferU << ":" << waferV << " index " << index;
+    }
+    if (verbosity_ > 1)
+      edm::LogVerbatim("HGCalValidation")
+          << "Detector " << name << " zside = " << zside << " layer = " << layer << " type = " << type << ":" << typef
+          << " wafer = " << waferU << ":" << waferV;
+  }
+}
+
+// ------------ method called when starting to processes a run  ------------
+void HGCalWaferHitCheck::beginRun(const edm::Run&, const edm::EventSetup& iSetup) {
+  edm::ESHandle<HGCalDDDConstants> pHGDC;
+  iSetup.get<IdealGeometryRecord>().get(nameDetector_, pHGDC);
+  hgcons_ = &(*pHGDC);
+  if (verbosity_ > 0)
+    edm::LogVerbatim("HGCalValidation") << nameDetector_ << " defined with " << hgcons_->layers(false)
+                                        << " Layers with " << (hgcons_->firstLayer() - 1) << " in front";
+
+  goodChannels_.clear();
+  for (int iz = 0; iz < 2; ++iz) {
+    int zside = iz * 2 - 1;
+    for (int layer = 1; layer <= layerMax; ++layer) {
+      for (int waferU = -waferMax; waferU <= waferMax; ++waferU) {
+        for (int waferV = -waferMax; waferV <= waferMax; ++waferV) {
+          int index = zside * HGCalWaferIndex::waferIndex(layer, waferU, waferV, false);
+          if (hgcons_->isValidHex8(layer, waferU, waferV, true))
+            goodChannels_.emplace_back(index);
+        }
+      }
+    }
+  }
+}
+
+void HGCalWaferHitCheck::endStream() {
```

Comment:
@makortel @Dr15Jones is modifying a `globalCache` in `endStream()` guaranteed to be atomic?

Reply:
Nope.

Reply:
That's what I was afraid of... @bsunanda this needs to be improved or redesigned in order to be thread-safe and avoid data races

Reply:
I am confused. I need to get the list of valid wafers from geometry - I can get it only in beginRun. How I can transfer that to globalCache? I am copying it to globalCache in beaginRun method. I understand it is done in beginRun - can I access geometry in beginJob or something equivalent which is called only on one thread

Reply:
> can I access geometry in beginJob 

EventSetup is not available in `beginJob()` or `endJob()` (or `beginStream()` or `endStream()`).

> How I can transfer that to globalCache?

One option is to protect the concurrent access to not-internally-thread-safe mutable members of the `globalCache()` e.g. with a spinlock or a mutex.